### ISSUE TITLE
Allow users to flag clean plugins

### ISSUE THREAD

**Issue by WrinklyNinja** [loot/loot#414]
This was [brought up](http://forums.bethsoft.com/topic/1513925-rel-loot-thread-17/?p=23920139) as a potentially helpful feature for users who go through their plugins and check which are clean, so that if they later forget, they don't need to go through and re-check.

A preliminary format would be a new key for the plugin map object, with its value being a list of clean plugin CRCs.
```
name: Example.esp
clean:
   - 0x01234567
```

If a plugin is clean, this could be communicated using either a message for the plugin or an icon.

The metadata could also get added to the masterlists and help other users too. The trouble with such a feature would be in ensuring cleanliness, especially regarding wild edits, which can be difficult to spot.

**Comment by Zone22** [comment]
It's similar to what I was thinking. The idea of adding a preliminary format would be a new key for the plugin map object, with its value being a list of clean plugin CRCs. With the primary purpose added to the masterlists and help other users. The idea really isn't about if a single end-user forgets and needs metadata to keep them from re-checking their own mods.
 
I think a little more discussion in light of my post regarding ensuring cleanliness as a matter of statistical benefit would be advantageous.
The ideas of the mod is clean as far as the current alliteration TES5Edit script can detect, and that there's appropriate channel of communication that's detailed enough for end-users who are experienced enough and care enough to report it when it's not clean.
 
This has several advantages clean is only by the definition TES5Edit automated script can detect. That means as TES5Edit matures so does the The accuracy of the clean mod being reported.
 
TES5Edit automated cleaning script is the definition in this case of cleanliness. Which we know is not perfect but is it good enough for the average user statistically speaking? The only way to ensure cleanliness is manually inspection. The end user would only do that if they have the experience or out of necessity due to issues or a report by the author. I imagine the dirty records that are not picked up by TES5Edit automated cleaning script are rare within themselves and even rarer that actually cause a problem within game. In which case if they do they could be reported.
 
Then there's the issue of abuse or user Error. If TES5Edit kept a small running record of plug-ins that had been checked with cleaning script LOOT could scrape that to determine whether or not the user has checked the records and that they are indeed as far as TES5Edit scripts can detect. Then and only then based on TES5Edit could the user report a plug-in is clean.

**Comment by WrinklyNinja** [comment]
> This has several advantages clean is only by the definition TES5Edit automated script can detect. That means as TES5Edit matures so does the The accuracy of the clean mod being reported.

Which also implies that earlier reports would then become of dubious validity unless they were checked, but having LOOT say they are clean would diminish the chance of that happening. At least a dirty mod is definitely dirty. Though it's not like there's any better check than TES5Edit we could use...

> I imagine the dirty records that are not picked up by TES5Edit automated cleaning script are rare within themselves and even rarer that actually cause a problem within game. In which case if they do they could be reported.

True. My counterpoints are splitting hairs a little, because any discrepancies are likely to be minor, as you say.

> Then there's the issue of abuse or user Error. If TES5Edit kept a small running record of plug-ins that had been checked with cleaning script LOOT could scrape that to determine whether or not the user has checked the records and that they are indeed clean as far as TES5Edit scripts can detect. Then and only then based on TES5Edit could the user report a plug-in as clean or dirty.

I assume you mean that record is a local file that TES5Edit writes to? That would help if there were an automated submission system, but I don't see it being helpful in the current situation of manual reports: if you were so inclined and going to the trouble of reporting it yourself, you'd also falsify any record.

---

This isn't going into 0.7 anyway, so I think it's worth sitting on and seeing what comes up in the time before the first post-0.7 release, but currently I'd be happy to implement a "Flag as clean" feature for users that only operates locally. I wouldn't be happy including such metadata into the masterlists though. Perhaps that will change with time.

**Comment by Freso** [comment]
I know I employ a comment syntax when adding `dirty` records to the masterlists, and I know at least some other editors do too. If `clean` (or whatever) ever makes it to the masterlist, I would probably encourage a similar system (as you've mentioned, a "clean" mod may actually be found to be dirty with a more recent *Edit).

Another thing to consider is "intentionally dirty" mods. Should they be flagged `clean` too (that is, if they don't *also* contain non-intentional dirty edits)?

**Comment by WrinklyNinja** [comment]
> Another thing to consider is "intentionally dirty" mods. Should they be flagged clean too (that is, if they don't also contain non-intentional dirty edits)?

I think there needs to be a difference between "you don't need to try cleaning this" and "you shouldn't clean any ITMs in this", so #323 will need a different solution.

After further thought, I'm still not comfortable with allowing for blanket "this is clean" metadata in the masterlist when this could later prove to be false. However, how about mirroring the `dirty` structures, eg.

```
name: Example.esp
clean:
   - crc: 0x01234567
     util: 'TES5Edit v3.1.0'
```

Then that metadata could be reported to the user as "Cleaned with TES5Edit v3.1.0", and the user could then decide if they wanted to double-check it was still clean with a more recent version of TES5Edit?

**Comment by Zone22** [comment]
I understand your hesitation but again those who have the skills and the aptitude would report "check" to verify clean plug-ins. Again those plug-ins that are reported clean that have issues are going to be evaluated.  In addition, if TES5Edit versioning is included in the metadata it would promote checking between versions for cleanliness. Although they should compared to the original not to the previously cleaned plug-in. That could be further reinforced through a warning in Loot. 

Think of many people instead of one user that will be checking their mods. Ultimately I think this is a good question to ask yourself. How are most plug-ins reported clean? Is it through using TES5Edit automated cleaning script or manual inspection by an author?  What has the best statistical benefit? 

In truth the decision lays in your hands. I'd love to hear from more people how this could be implemented in a way that promotes mod cleanliness and end-user responsibility.

**Comment by WrinklyNinja** [comment]
@LOOT, @Zone22: I'd like to get this implemented for v0.10.0 - any issues with the syntax being as I suggested above, ie.

```
name: Example.esp
clean:
   - crc: 0x01234567
     util: 'TES5Edit v3.1.0'
```

? The `crc` and `util` fields would both be mandatory. The user would then see a message or icon with a tooltip that reads something like "Cleaned with TES5Edit v3.1.0", saving them unnecessary effort or, if they wanted to, letting them know it could be re-checked with a more recent release of TES5Edit.

**Comment by Zone22** [comment]
The fields are satisfactory. Is the metadata going to be kept a local or on the master list?

**Comment by WrinklyNinja** [comment]
I'm happy with it being stored in either/both.

**Comment by Zone22** [comment]
Having there's benefits of having it to being both so that's just my two cents.

**Comment by WrinklyNinja** [comment]
OK, how's this UI for showing that a plugin is clean (ignore that it's for Skyrim.esm):

![screenshot 49](https://cloud.githubusercontent.com/assets/2430738/16547205/3fd7ec84-415d-11e6-92fe-6435f662ff8b.png)

I have trouble choosing appropriate icons, so if you've got a suggestion for a different icon that would be good too.

**Comment by WrinklyNinja** [comment]
I decided to try making a brush icon, is this better?

![screenshot 50](https://cloud.githubusercontent.com/assets/2430738/16567960/6f5be41e-421c-11e6-902a-d490e9803c95.png)

(Pixel art is hard...)

**Comment by WrinklyNinja** [comment]
This is now done in the `masterlist-syntax-changes` branch, but I'll leave the issue open until the branch is merged.

**Comment by Freso** [comment]
"Cleaned with …" seems misleading. Isn't the point of the `clean` flag that it doesn't need cleaning, rather than it already having been cleaned?

Also, should entries be added to the masterlist now/on the `v0.10` branch for plugins that do not otherwise have any information, but have been checked and found to be clean?

**Comment by Freso** [comment]
(Also not a huge fan of the brush icon, but haven't got a better suggestion. :))

**Comment by WrinklyNinja** [comment]
> "Cleaned with …" seems misleading. Isn't the point of the clean flag that it doesn't need cleaning, rather than it already having been cleaned?

Well, both. How about "Verified clean with ..."?

> Also, should entries be added to the masterlist now/on the v0.10 branch for plugins that do not otherwise have any information, but have been checked and found to be clean?

You can add entries now to the `v0.8` branch, since `clean` is a new key it won't cause compatibility issues.

**Comment by WrinklyNinja** [comment]
> (Also not a huge fan of the brush icon, but haven't got a better suggestion. :))

Well, what don't you like about it? Maybe we can come up with a better idea. I tried a brush because that's what I get when I do an image search for "clean icon".

**Comment by pStyl3** [comment]
The brush+tick icon is the best candidate for it's purpose to date, but it's still a combination of two icons, which seems a bit "arbitrary", plus .. LOOT already uses the tick icon to indicate the active/inactive state of a plugin. I would think that Freso's concerns may come from this direction.

Try searching for "pure icon", there are some interesting icons like [this one](http://st2.depositphotos.com/5205783/7954/v/450/depositphotos_79545632-Logo-water-drop-pure-symbol.jpg) aswell, so some sort of "water drop"-icon. Cleanliness is often associated with clean water, so it might be worth a try to go in that direction.

**Comment by WrinklyNinja** [comment]
> Try searching for "pure icon", there are some interesting icons like this one aswell, so some sort of "water drop"-icon. Cleanliness is often associated with clean water, so it might be worth a try to go in that direction.

Good luck getting that across in 24x24 black and white pixels. :stuck_out_tongue_closed_eyes: Well, it's a compound icon because people said that just having a brush is ambigous: does it need cleaning or is it already clean? I can't think of an alternative icon where that isn't the case...

**Comment by Zone22** [comment]
@Freso 
"Cleaned with …" seems misleading."
If Loot says the plug-in is as clean then as it's as clean going to be via TES5Edit automated cleaning script at the given TES5Edit version number.

There's only two instances when someone would need to check if the plug-ins clean thats flag by Loot.
1. New version of TES5Edit. That's why "Cleaned with TES5Edit V X..X.X" If the user has a higher version number then what's found in the metadata they should check if the plug-ins clean. 

2. If they are having problems in game. They may need to cleanup a plug-in manually via TES5Edit.

**Comment by WrinklyNinja** [comment]
> If Loot says the plug-in is as clean then as it's as clean going to be via TES5Edit automated cleaning script at the given TES5Edit version number.

I think the misleading bit is that a plugin may not have needed cleaning in the first place.

**Comment by Freso** [comment]
My issue with the brush is that brush makes me think of painting (painting something over?), not of cleaning or being clean. The clipboard with a tick makes more sense to me.

> I think the misleading bit is that a plugin may not have needed cleaning in the first place.

Yep.

**Comment by Zone22** [comment]
I really have no preference on the icon. Otherwise I feel like I'm missing something you're both trying to communicate... I apologize for my lack of perception...
> I think the misleading bit is that a plugin may not have needed cleaning in the first place.

Unless an ESP file has been processed manually or automatically an individual has no idea whether or not it's clean. In addition I was under the impression part of the reasoning of including "clean" metadata as allow the end-user to save time by not having to check the cleanliness all their mods.  I'm still a little lost on what could be construed as misleading. Could you explain a little bit more?

The only misunderstanding I can first see if the end-user believes a mod is clean with an older version TES5Edit.  Therefore the metadata is never updated because someone thinks of the mod is already clean. 

In order to avoid this confusion the UI needs to flag the user to check if a mod is clean or not based on the current version of TES5Edit?

**Comment by WrinklyNinja** [comment]
> Could you explain a little bit more?

If a mod author creates a plugin, checks it for dirty edits in TES5Edits and finds none, the plugin is clean without having been cleaned, so LOOT telling users that the plugin was cleaned is incorrect.

**Comment by WrinklyNinja** [comment]
> My issue with the brush is that brush makes me think of painting

It's not a paintbrush though!

**Comment by Zone22** [comment]
> If a mod author creates a plugin, checks it for dirty edits in TES5Edits and finds none, the plugin is clean without having been cleaned, so LOOT telling users that the plugin was cleaned is incorrect.

So the issue is clean versus cleaned. In my mind I had it framed that a plug-in is a known state which consists of "dirty" or "clean" to be a known "state" of a Mod. I now see how the wording can be considered misleading. The current wording suggests an implied process to achieve the "clean" state.

> TES5Edit automated cleaning script is the definition in this case of cleanliness
if we operate off of that definition we could say "_Verified_ with..."

**Comment by pStyl3** [comment]
What about something like [this](https://www.iconfinder.com/icons/464472/clean_drop_ecology_water_icon#size=128) .. that's a very simplistic, yet stylish icon.

**Comment by WrinklyNinja** [comment]
Does anyone have any criticism against the message bring "Verified clean by..."?

The use of "by" instead of "with" is to distinguish that it uses "clean" as an adjective, not a verb, so users don't think the message means that the plugin has been verified to need cleaning with the given utility.

I'm also fine with using a water drop icon instead of the brush + checkmark, @Freso, any complaint from you? At the end of the day, the icon needs to be associable, but not necessarily intuitive (because it's a lot harder to get right).

**Comment by Zone22** [comment]
> Does anyone have any criticism against the message bring "Verified clean by..."?
Not I

**Comment by WrinklyNinja** [comment]
Merged in f6512c6d35c66dbf80432589235eb9803d6871eb
